internal/dataplane/kongstate/customentity.go

```diff
@@ -140,6 +147,102 @@ type SchemaGetter interface {
 	Get(ctx context.Context, entityType string) (kong.Schema, error)
 }
 
+type entityForeignFieldValue struct {
+	fieldName         string
+	foreignEntityType kong.EntityType
+	foreignEntityID   string
+}
+
+// FillCustomEntities fills custom entities in KongState.
+func (ks *KongState) FillCustomEntities(
+	logger logr.Logger,
+	s store.Storer,
+	failuresCollector *failures.ResourceFailuresCollector,
+	schemaGetter SchemaGetter,
+	workspace string,
+) {
+	entities := s.ListKongCustomEntities()
+	if len(entities) == 0 {
+		return
+	}
+	logger = logger.WithName("fillCustomEntities")
+
+	if ks.CustomEntities == nil {
+		ks.CustomEntities = map[string]*KongCustomEntityCollection{}
+	}
+	// Fetch relations between plugins and services/routes/consumers and store the pointer to translated Kong entities.
+	// Used for fetching entity referred by a custom entity and fill the ID of referred entity.
+	pluginRels := ks.getPluginRelatedEntitiesRef(s, logger)
+
+	for _, entity := range entities {
+		// reject the custom entity if its type is in "known" entity types that are already processed.
+		if IsKnownEntityType(entity.Spec.EntityType) {
+			failuresCollector.PushResourceFailure(
+				fmt.Sprintf("cannot use known entity type %s in custom entity", entity.Spec.EntityType),
+				entity,
+			)
+			continue
+		}
+		// Fetch the entity schema.
+		schema, err := ks.fetchEntitySchema(schemaGetter, entity.Spec.EntityType)
+		if err != nil {
+			failuresCollector.PushResourceFailure(
+				fmt.Sprintf("failed to fetch entity schema for entity type %s: %v", entity.Spec.EntityType, err),
+				entity,
+			)
+			continue
+		}
+
+		// Unmarshal fields of the entity.
+		var parsedEntity map[string]any
+		if err = json.Unmarshal(entity.Spec.Fields.Raw, &parsedEntity); err != nil {
+			failuresCollector.PushResourceFailure(fmt.Sprintf("failed to unmarshal fields of entity: %v", err), entity)
+			continue
+		}
```

Comment:
I think we can skip this as it is done in `generateCustomEntities`.
```suggestion
```

Reply:
Done.